crates/pico/src/memo.rs

```diff
@@ -0,0 +1,188 @@
+        time_calculated
+    };
+    register_dependency(
+        db,
+        NodeKind::Derived(node_id),
+        time_calculated,
+        current_epoch,
+    );
+    state
+}
+
+fn any_dependency_changed<Db: Database>(
+    db: &mut Db,
+    node_id: &DerivedNodeId,
+    current_epoch: Epoch,
+) -> bool {
+    let dependencies = db
+        .storage()
+        .derived_nodes()
+        .get(node_id)
+        .expect("node should exist. This is indicative of a bug in Pico.")
+        .dependencies
+        .iter()
+        .filter_map(|dep| {
+            if dep.time_verified_or_calculated != current_epoch {
+                Some((dep.node_to, dep.time_verified_or_calculated))
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<_>>();
+
+    dependencies
+        .into_iter()
+        .any(|(dep_node, time_verified)| match dep_node {
+            NodeKind::Source(key) => source_node_changed(db, &key, time_verified),
+            NodeKind::Derived(dep_node_id) => derived_node_changed(db, dep_node_id),
+        })
+}
+
+fn source_node_changed<Db: Database>(db: &Db, key: &Key, time_verified: Epoch) -> bool {
+    match db.storage().source_nodes().get(key) {
+        Some(source) => source.time_calculated > time_verified,
+        None => true,
+    }
+}
+
+fn derived_node_changed<Db: Database>(db: &mut Db, node_id: DerivedNodeId) -> bool {
+    let inner_fn = if let Some(node) = db.storage().derived_nodes().get(&node_id) {
+        node.inner_fn
+    } else {
+        return true;
+    };
+    if !db.storage().params().contains_key(&node_id.param_id) {
+        return true;
+    }
+    let state = memo(db, node_id, inner_fn);
+    matches!(state, MemoState::Computed)
+}
+
+fn call_inner_fn_and_collect_dependencies<Db: Database>(
+    db: &mut Db,
+    param_id: ParamId,
+    inner_fn: fn(&mut Db, ParamId) -> Box<dyn DynEq>,
+) -> (
+    Box<dyn DynEq>,  /* value */
+    Vec<Dependency>, /* dependencies */
+    Epoch,           /* time_calculated */
+) {
+    let (value, registered_dependencies) = with_dependency_tracking(db, param_id, inner_fn);
+    let (dependencies, time_calculated) = registered_dependencies.into_iter().fold(
+        (vec![], Epoch::new()),
+        |(mut deps, mut max_time_calculated), (time_calculated, dep)| {
+            deps.push(dep);
+            max_time_calculated = std::cmp::max(max_time_calculated, time_calculated);
+            (deps, max_time_calculated)
+        },
+    );
+    (value, dependencies, time_calculated)
+}
+
+pub fn register_dependency<Db: Database>(
+    db: &mut Db,
+    node: NodeKind,
+    time_calculated: Epoch,
+    current_epoch: Epoch,
+) {
+    if let Some(dependencies) = db.storage_mut().dependency_stack().last_mut() {
+        dependencies.push((
+            time_calculated,
+            Dependency {
+                node_to: node,
+                time_verified_or_calculated: current_epoch,
+            },
+        ));
+    }
```

Comment:
can you add an `else { // do nothing. We just finished the outermost memoized function }` or something for clarity?